Enable 64bit `long double`

Blackhex opened the pull request:
- I'd split enabling 64bit `long double` and enabling `_Float128` support into separate commits/PRs.
- If `long double` size is not 128bit, libgfortran requires having `_Float32`. For this reason, 10 and 16 byte `REAL` has been disabled in this PR until `_Float32` support will be added in the next PR.

Closes https://github.com/ZacWalk/mingw-woarm64-build/issues/9

Blackhex committed:
Enable 64-bit long double

Blackhex committed:
Disable libgfortran 10 and 16 byte REAL - it requires either _Float128 or 10 and 16 byte long double

Blackhex commented:
@ZacWalk In the context of the above, is `_Float128` actually needed for something?

ZacWalk commented:
If _Float128  works now I believe we should preserve that functionality.

The main issue I see with these changes is that they need to be specific to aarch64-coff. All other build targets should be unchanged. 

For example, aarch64.cc is a file also used by many aarch64 targets. Code changes in this file need to be coff specific - maybe behind an ifdef.

Blackhex commented:
> If \_Float128 works now I believe we should preserve that functionality.

`_Float128` is supported only for some targets. Whether it should be supported on `aarch64-w64-mingw32` is a question to find out. I'd say it should be suported only if it is supported both on `aarch64-pc-linux` and `x86_64-w64-mingw32`.

eukarpov commented:
here is a link to CI check
https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/actions/runs/7007818760

eukarpov commented on gcc/config/aarch64/aarch64-coff.h:
```diff
@@ -34,9 +34,14 @@
 #define PTRDIFF_TYPE	"long long int"
 
 #define TARGET_64BIT 1
+#define TARGET_LONG_DOUBLE_64 1
```
Why is TARGET_LONG_DOUBLE_64 used instead of TARGET_LONG_DOUBLE_128?
Do we need to support 64 long double for aarch64?
why it is not used by aarch64 for other platforms and object formats?

Blackhex replied:
64bit double is implemented in hardware, 128bit double must be emulated. MSVC uses 64bit double.

vejbomar replied:
More thoroughly discussed in https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/issues/9

eukarpov commented:
Let's postpone review of this PR for a while as it was discussed

vejbomar commented on gcc/config/aarch64/aarch64-coff.h:
```diff
@@ -31,9 +31,15 @@
 #define LONG_TYPE_SIZE 32
 
+#undef LONG_DOUBLE_TYPE_SIZE
+#define LONG_DOUBLE_TYPE_SIZE 64
```
The way how floating point type sizes are defined was refactored in https://github.com/Windows-on-ARM-Experiments/gcc-woarm64/commit/55947b32c38a40777aedbd105bd94b43a42c2a10 which also marked `LONG_DOUBLE_TYPE_SIZE` as poisoned. (Failed build log: https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/actions/runs/12201632307/job/34040460962)

vejbomar committed:
Fix enabling 64-bit long double

Fixed after refactor in upstream https://gcc.gnu.org/pipermail/gcc-patches/2024-June/653338.html

vejbomar commented on libgfortran/configure.ac:
```diff
@@ -220,7 +220,7 @@ AM_CONDITIONAL(LIBGFOR_MINIMAL, false)
 # although they do have the support.
 LIBGOMP_CHECKED_INT_KINDS="1 2 4 8 16"
-LIBGOMP_CHECKED_REAL_KINDS="4 8 10 16"
+LIBGOMP_CHECKED_REAL_KINDS="4 8"
```
@Blackhex do you remember why was this was changed to `"4 8"` for all platforms and not just for `aarch64-*-mingw*` like it was done in the same [commit](https://github.com/Windows-on-ARM-Experiments/gcc-woarm64/pull/3/commits/3715b4f1e8f106e352f2ba09cf826d8dbf3893b0) in `libfortran/configure`?

Blackhex replied:
I was just lazy. This needs to be conditioned.

vejbomar commented:
Changes verified in https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/actions/runs/12596640185

Blackhex commented on gcc/config/aarch64/aarch64.cc:
```diff
@@ -29090,10 +29090,12 @@ aarch64_scalar_mode_supported_p (scalar_mode mode)
-	  : default_scalar_mode_supported_p (mode));
+  else if (mode == TFmode)
+    return true;
+  else if (mode == HFmode || mode == BFmode)
```
Why not merge those conditions? But in general, I am not sure this change is correct.

vejbomar replied:
I changed the code specific for woarm64 to be guarded by `ifdef`, however, the question if this change is correct is still on the table.

As I understand it, we have two options:
1. Have this code present which declares that woarm64 target is prepared to handle the 128bit floats ([doc string](https://github.com/Windows-on-ARM-Experiments/gcc-woarm64/blob/e368cd1114b56cda876c1ac58812a2f599b6213f/gcc/doc/tm.texi#L4582)). Afaik woarm64 target is prepared to handle 128 bit floats, there just isn't currently any way how to define them in C++ code on woarm64. See aarch64 code defines [support for 128bit float mode](https://github.com/Windows-on-ARM-Experiments/gcc-woarm64/blob/e368cd1114b56cda876c1ac58812a2f599b6213f/gcc/config/aarch64/aarch64-modes.def#L261) and define [software emulation for 128bit operations](https://github.com/Windows-on-ARM-Experiments/gcc-woarm64/blob/e368cd1114b56cda876c1ac58812a2f599b6213f/libgcc/config/aarch64/t-softfp#L1).
2. Guard the code which defines that 128bit floats are supported on aarch64 to not be used on woarm64. Which needs changes at least at the two places linked above and maybe more. This would be a bit more complicated to implement and would be reverted again when adding support for `_float128`

What do you think? I would go for 1. because it is faster and prepared for having `_float128` or `std::float128_t` on woarm64 later. Not sure thought what would be the stance of maintainers.

Blackhex replied:
I don't have any preference to do either of those options. The situation was the same when I was working on it a few months ago: Either introduce just the `long long double` changes and disable 128-bit float support or do both at the same time which means fixing `gfortran`. There is no other immediate justification for focusing on 128-bit float support though.

vejbomar replied:
Ok, I agree that adding support for 128-bit float is not a priority now, so I won't work on it now. But sticking to this code change as the simplest way how to implement 64bit long double and keeping code prepared for 128bit floating point type.

vejbomar committed:
Limit checked float sizes only for Aarch64 MinGW

eukarpov commented on gcc/config/aarch64/aarch64-coff.h:
```diff
@@ -31,9 +31,14 @@
 #undef PTRDIFF_TYPE
 #define PTRDIFF_TYPE	"long long int"
 
+#define TARGET_64BIT 1
```
TARGET_64BIT is coming from ix86 and should be replaced if needed with TARGET_AARCH64_MS_ABI

Blackhex replied:
That's not completely true. `TARGET_64BIT` is used throughout the codebase for multiple targets and its semantics is cross-target.  I agree that defining it here and now might not be needed. It depends on the context for what code is this define is needed? Unless AArch64 maintainers said explicitly otherwise...

eukarpov replied:
TARGET_64BIT has been removed from all patch series which have been upstreamed so far based on community feedback.

vejbomar replied:
Removed

vejbomar committed:
Specify that woarm64 target is prepared to handle 128bit floating point operations

vejbomar committed:
Remove unnecessary definition

eukarpov commented:
LGTM to be merged internally. Please introduce a test for this change in mingw-build internal tests.
More suggestions will be received from the community during the upstreaming.

vejbomar commented:
"Build toolchain variants" run for current state of this PR: https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/actions/runs/12692740574

Blackhex commented on gcc/config/aarch64/aarch64-coff.h:
```diff
@@ -31,9 +31,13 @@
 #undef LONG_TYPE_SIZE
 #define LONG_TYPE_SIZE 32
 
+#define __NO_BINARY80__
```
I think this should not be needed. The only usage of this macro, I have found is at https://github.com/Windows-on-ARM-Experiments/gcc-woarm64/blob/woarm64/libgcc/config/libbid/bid_functions.h#L70. Unless `__INTEL_COMPILER` macro is defined (which shouldn't be),  it evaluates to `__ENABLE_BINARY80__` being undefined while the `__ENABLE_BINARY80__` is always checked as `#if __ENABLE_BINARY80__`.

vejbomar replied:
It was needed, if I'm not mistaken `__ENABLE_BINARY80__` is defined because `!defined _MSC_VER` will pass in the code you send https://github.com/Windows-on-ARM-Experiments/gcc-woarm64/blob/e5c582536a57cbe031770db1515c41e7aa3a11cd/libgcc/config/libbid/bid_functions.h#L73C5-L73C22

Blackhex replied:
You are right, I missed the `||` condition. On the other hand, it might be better to handle the case at https://github.com/Windows-on-ARM-Experiments/gcc-woarm64/blob/woarm64/libgcc/config/libbid/bid_functions.h#L70 instead of in `aarch64-coff.h`. We will see what the reviewers will prefer. For now, I am fine with the current state.

Blackhex commented on libgfortran/configure.ac:
```diff
@@ -220,7 +220,12 @@ AM_CONDITIONAL(LIBGFOR_MINIMAL, false)
 LIBGOMP_CHECKED_INT_KINDS="1 2 4 8 16"
-LIBGOMP_CHECKED_REAL_KINDS="4 8 10 16"
+case "${host}" in
+  aarch64-*-mingw*)
```
Did you test if this does not break `aarch64-pc-cygwin` build?

vejbomar replied:
Good point, build of `aarch64-pc-cygwin` is failing with these changes applied. From a quick look, I don't think it's caused by this change but some more changes will be needed in this PR.

https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/actions/runs/12763967760/job/35575173609

vejbomar replied:
Fixed in https://github.com/Windows-on-ARM-Experiments/newlib-cygwin/pull/6 (which needs to be merged first) and verified by https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/actions/runs/12912580048

Blackhex commented on gcc/config/aarch64/aarch64.cc:
```diff
@@ -29109,6 +29109,15 @@ aarch64_scalar_mode_supported_p (scalar_mode mode)
+     port is prepared to handle TFmode instructions. If long double is
+     128bit, this is handled by default_scalar_mode_supported_p.  */
+
+  #ifdef TARGET_LONG_DOUBLE_64
```
Should not be indented.

Blackhex reviewed (commented):
Looks good in general, left some minor coments.

vejbomar committed:
Disable libfortran 10 and 16 byte REAL also for aarch64-pc-cygwin

eukarpov commented on libgfortran/configure.ac:
```diff
@@ -220,7 +220,12 @@ AM_CONDITIONAL(LIBGFOR_MINIMAL, false)
 LIBGOMP_CHECKED_INT_KINDS="1 2 4 8 16"
-LIBGOMP_CHECKED_REAL_KINDS="4 8 10 16"
+case "${host}" in
+  aarch64-*-mingw*)
```
Should it be also adjusted for cygwin?

vejbomar replied:
Good point, thanks! It was accidentally removed in the last rebase.

vejbomar commented:
Validated by https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/actions/runs/13014388392
